azure-iot-device/azure/iot/device/common/pipeline/pipeline_thread.py

```diff
@@ -0,0 +1,199 @@
+    return _get_thread_local_storage.local
+
+
+def _invoke_on_executor_thread(thread_name, block=True, _func=None):
+    """
+    Decorator which runs the wrapped function on a given thread.  If block==False,
+    the call returns immediately without waiting for the decorated function to complete.
+    If block==True, the call waits for the decorated function to complete before returning.
+    """
+
+    def decorator(func):
+        function_name = getattr(func, "__name__", str(func))
+
+        def wrapper(*args, **kwargs):
+            if not getattr(_get_thread_local_storage(), "in_{}_thread".format(thread_name), False):
+                logger.info("Starting {} in {} thread".format(function_name, thread_name))
+
+                def thread_proc():
+                    setattr(_get_thread_local_storage(), "in_{}_thread".format(thread_name), True)
+                    return func(*args, **kwargs)
+
+                # TODO: add a timeout here and throw exception on failure
+                future = _get_named_executor(thread_name).submit(thread_proc)
+                if block:
+                    return future.result()
+                else:
+                    return future
+            else:
+                logger.debug("Already in {} thread for {}".format(thread_name, function_name))
+                return func(*args, **kwargs)
+
+        # Silly hack:  On 2.7, we can't use @functools.wraps on callables don't have a __name__ attribute
+        # attribute(like MagicMock object), so we only do it when we have a name.  functools.update_wrapper
+        # below is the same as using the @functools.wraps(func) decorator on the wrapper function above.
+        if getattr(func, "__name__", None):
+            return functools.update_wrapper(wrapped=func, wrapper=wrapper)
+        else:
+            return wrapper
+
+    if _func is None:
+        return decorator
+    else:
+        return decorator(_func)
+
+
+def invoke_on_pipeline_thread(_func=None):
+    """
+    Run the decorated function on the pipeline thread.
+    """
+    return _invoke_on_executor_thread(thread_name="pipeline", _func=_func)
+
+
+def invoke_on_pipeline_thread_nowait(_func=None):
+    """
+    Run the decorated function on the pipeline thread, but don't wait for it to complete
+    """
+    return _invoke_on_executor_thread(thread_name="pipeline", block=False, _func=_func)
+
+
+def invoke_on_callback_thread_nowait(_func=None):
+    """
+    Run the decorated function on the callback thread, but don't wait for it to complete
+    """
+    return _invoke_on_executor_thread(thread_name="callback", block=False, _func=_func)
+
+
+def _assert_executor_thread(thread_name, _func=None):
+    """
+    Decorator which asserts that the given function only gets called inside the given
+    thread.
+    """
+
+    def decorator(func):
+        function_name = getattr(func, "__name__", str(func))
+
+        @functools.wraps(func)
```

Comment:
>@functools.wraps(func) [](start = 8, length = 22)

Is this not going to have the same issue as above?

Reply:
yes same question i had

---
In reply to: [301796343](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r301796343) [](ancestors = 301796343)

Reply:
We only need that hack above because we use this decorator to wrap MicroMock objects

---
In reply to: [301843801](https://github.com/Azure/azure-iot-sdk-python-preview/pull/133#discussion_r301843801) [](ancestors = 301843801,301796343)